resolver/bootstrap.go

```diff
@@ -29,56 +29,58 @@ type Bootstrap struct {
 	log *logrus.Entry
 
 	resolver    Resolver
-	upstream    Resolver // the upstream that's part of the above resolver
-	upstreamIPs []net.IP // IPs for b.upstream
+	bootstraped bootstrapedResolvers
 
 	systemResolver *net.Resolver
 }
 
 // NewBootstrap creates and returns a new Bootstrap.
 // Internally, it uses a CachingResolver and an UpstreamResolver.
 func NewBootstrap(cfg *config.Config) (b *Bootstrap, err error) {
-	upstream := cfg.BootstrapDNS.Upstream
 	log := log.PrefixedLog("bootstrap")
 
-	var ips []net.IP
-
-	switch {
-	case upstream.IsDefault():
-		log.Infof("bootstrapDns is not configured, will use system resolver")
-	case upstream.Net == config.NetProtocolTcpUdp:
-		ip := net.ParseIP(upstream.Host)
-		if ip == nil {
-			return nil, fmt.Errorf("bootstrapDns uses %s but is not an IP", upstream.Net)
-		}
-
-		ips = append(ips, ip)
-	default:
-		ips = cfg.BootstrapDNS.IPs
-		if len(ips) == 0 {
-			return nil, fmt.Errorf("bootstrapDns.IPs is required when upstream uses %s", upstream.Net)
-		}
-	}
-
 	// Create b in multiple steps: Bootstrap and UpstreamResolver have a cyclic dependency
 	// This also prevents the GC to clean up these two structs, but is not currently an
 	// issue since they stay allocated until the process terminates
 	b = &Bootstrap{
 		log:            log,
-		upstreamIPs:    ips,
 		systemResolver: net.DefaultResolver, // allow replacing it during tests
 	}
 
-	if upstream.IsDefault() {
+	bootstraped, err := newBootstrapedResolvers(b, cfg.BootstrapDNS)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(bootstraped) == 0 {
+		log.Infof("bootstrapDns is not configured, will use system resolver")
+
 		return b, nil
 	}
 
-	b.upstream = newUpstreamResolverUnchecked(upstream, b)
+	parallelResolver, err := newParallelBestResolver(bootstraped.ResolverGroups())
+	if err != nil {
+		return nil, fmt.Errorf("could not create bootstrap ParallelBestResolver: %w", err)
+	}
+
+	// Always enable prefetching to avoid stalling user requests
+	// Otherwise, a request to blocky could end up waiting for 2 DNS requests:
+	//   1. lookup the DNS server IP
+	//   2. forward the user request to the server looked-up in 1
+	cachingCfg := cfg.Caching
+	cachingCfg.EnablePrefetch()
+
+	if cachingCfg.MinCachingTime == 0 {
+		// Set a min time in case the user didn't to avoid prefetching too often
+		cachingCfg.MinCachingTime = config.Duration(time.Hour)
+	}
```

Comment:
Testing this some more, I noticed that `caching.minTime` has no default.  
Since with this MR, the caching resolver is always enabled with prefetching as part of the `Bootstrap`, this can cause a lot of churn for not much gains.
I noticed because the domain `s3.amazonaws.com` used by the list `https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt` I have in my test config has a TTL of 3, and was being prefetched every 10 seconds.

I added a default of 1h here, only if the user hasn't specified anything (or they put 0). Let me know if you think I should revert this and we should just trust the TTLs.